{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyOfovLd+FeofuXn7SMh6TNf",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/BayarBerkeley/DummyProject/blob/main/lycopene_pathway.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 328,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7Zrwnz3yMEWX",
        "outputId": "e2422e8d-2705-45ea-cdb3-6859c5049be7"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n",
            "Requirement already satisfied: cobra in /usr/local/lib/python3.9/dist-packages (0.26.3)\n",
            "Requirement already satisfied: optlang~=1.5 in /usr/local/lib/python3.9/dist-packages (from cobra) (1.6.1)\n",
            "Requirement already satisfied: depinfo~=1.7 in /usr/local/lib/python3.9/dist-packages (from cobra) (1.7.0)\n",
            "Requirement already satisfied: diskcache~=5.0 in /usr/local/lib/python3.9/dist-packages (from cobra) (5.5.1)\n",
            "Requirement already satisfied: swiglpk in /usr/local/lib/python3.9/dist-packages (from cobra) (5.0.8)\n",
            "Requirement already satisfied: ruamel.yaml~=0.16 in /usr/local/lib/python3.9/dist-packages (from cobra) (0.17.21)\n",
            "Requirement already satisfied: httpx~=0.24 in /usr/local/lib/python3.9/dist-packages (from cobra) (0.24.0)\n",
            "Requirement already satisfied: numpy<1.24,>=1.13 in /usr/local/lib/python3.9/dist-packages (from cobra) (1.22.4)\n",
            "Requirement already satisfied: appdirs~=1.4 in /usr/local/lib/python3.9/dist-packages (from cobra) (1.4.4)\n",
            "Requirement already satisfied: importlib-resources in /usr/local/lib/python3.9/dist-packages (from cobra) (5.12.0)\n",
            "Requirement already satisfied: rich>=8.0 in /usr/local/lib/python3.9/dist-packages (from cobra) (13.3.3)\n",
            "Requirement already satisfied: pydantic~=1.6 in /usr/local/lib/python3.9/dist-packages (from cobra) (1.10.7)\n",
            "Requirement already satisfied: future in /usr/local/lib/python3.9/dist-packages (from cobra) (0.18.3)\n",
            "Requirement already satisfied: python-libsbml~=5.19 in /usr/local/lib/python3.9/dist-packages (from cobra) (5.20.0)\n",
            "Requirement already satisfied: pandas~=1.0 in /usr/local/lib/python3.9/dist-packages (from cobra) (1.5.3)\n",
            "Requirement already satisfied: certifi in /usr/local/lib/python3.9/dist-packages (from httpx~=0.24->cobra) (2022.12.7)\n",
            "Requirement already satisfied: httpcore<0.18.0,>=0.15.0 in /usr/local/lib/python3.9/dist-packages (from httpx~=0.24->cobra) (0.17.0)\n",
            "Requirement already satisfied: sniffio in /usr/local/lib/python3.9/dist-packages (from httpx~=0.24->cobra) (1.3.0)\n",
            "Requirement already satisfied: idna in /usr/local/lib/python3.9/dist-packages (from httpx~=0.24->cobra) (3.4)\n",
            "Requirement already satisfied: sympy>=1.0 in /usr/local/lib/python3.9/dist-packages (from optlang~=1.5->cobra) (1.11.1)\n",
            "Requirement already satisfied: six>=1.9 in /usr/local/lib/python3.9/dist-packages (from optlang~=1.5->cobra) (1.16.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.1 in /usr/local/lib/python3.9/dist-packages (from pandas~=1.0->cobra) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.9/dist-packages (from pandas~=1.0->cobra) (2022.7.1)\n",
            "Requirement already satisfied: typing-extensions>=4.2.0 in /usr/local/lib/python3.9/dist-packages (from pydantic~=1.6->cobra) (4.5.0)\n",
            "Requirement already satisfied: markdown-it-py<3.0.0,>=2.2.0 in /usr/local/lib/python3.9/dist-packages (from rich>=8.0->cobra) (2.2.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.9/dist-packages (from rich>=8.0->cobra) (2.14.0)\n",
            "Requirement already satisfied: ruamel.yaml.clib>=0.2.6 in /usr/local/lib/python3.9/dist-packages (from ruamel.yaml~=0.16->cobra) (0.2.7)\n",
            "Requirement already satisfied: zipp>=3.1.0 in /usr/local/lib/python3.9/dist-packages (from importlib-resources->cobra) (3.15.0)\n",
            "Requirement already satisfied: h11<0.15,>=0.13 in /usr/local/lib/python3.9/dist-packages (from httpcore<0.18.0,>=0.15.0->httpx~=0.24->cobra) (0.14.0)\n",
            "Requirement already satisfied: anyio<5.0,>=3.0 in /usr/local/lib/python3.9/dist-packages (from httpcore<0.18.0,>=0.15.0->httpx~=0.24->cobra) (3.6.2)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.9/dist-packages (from markdown-it-py<3.0.0,>=2.2.0->rich>=8.0->cobra) (0.1.2)\n",
            "Requirement already satisfied: mpmath>=0.19 in /usr/local/lib/python3.9/dist-packages (from sympy>=1.0->optlang~=1.5->cobra) (1.3.0)\n"
          ]
        }
      ],
      "source": [
        "!pip install cobra"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Here, we extract genomic, metabolic, and reaction data of Escherichia coli str. K-12 substr. MG1655 from the BiGG database using <i>load_model"
      ],
      "metadata": {
        "id": "fPQzLkiWfT_4"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from cobra.io import load_model\n",
        "import pandas as pd\n",
        "\n",
        "# load Escherichia coli str. K-12 substr. MG1655 from BiGG database\n",
        "ecoli_model = load_model(\"iJO1366\")\n",
        "print(\"Size of metabolites: \", len(ecoli_model.metabolites))\n",
        "print(\"Size of reactions: \", len(ecoli_model.reactions))\n",
        "print(\"Size of genes: \", len(ecoli_model.genes))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Qch56vayMRLO",
        "outputId": "ca95da62-ed5e-4608-81fd-606e44607ae6"
      },
      "execution_count": 329,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Size of metabolites:  1805\n",
            "Size of reactions:  2583\n",
            "Size of genes:  1367\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Using the subsystem of the reaction, we can extract the Glycolysis pathway reaction from our <i>ecoli_model."
      ],
      "metadata": {
        "id": "FaTJaWynPa6l"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Get the glycolysis subsystem reactions\n",
        "glycolysisSubsystem = \"Glycolysis/Gluconeogenesis\"\n",
        "glycolysisReactions = [r for r in ecoli_model.reactions if r.subsystem == glycolysisSubsystem]\n",
        "\n",
        "# Create a table for the reactions\n",
        "reactionNames = [r.name for r in glycolysisReactions]\n",
        "reactionFormulas = [r.build_reaction_string() for r in glycolysisReactions]\n",
        "tableData = {'Reaction_Names': reactionNames, 'Reaction_Formulas': reactionFormulas}\n",
        "tableIndex = [r.id for r in glycolysisReactions]\n",
        "T = pd.DataFrame(tableData, index=tableIndex)\n",
        "T"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 739
        },
        "id": "Iv86cnRCM9Yx",
        "outputId": "9ed05c2c-0243-40e0-f107-02f50aab31c8"
      },
      "execution_count": 330,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                             Reaction_Names  \\\n",
              "ENO                                                 Enolase   \n",
              "F6PA                          Fructose 6-phosphate aldolase   \n",
              "FBA                          Fructose-bisphosphate aldolase   \n",
              "FBP                                 Fructose-bisphosphatase   \n",
              "G1PPpp                                Glucose-1-phosphatase   \n",
              "G6PP                        Glucose-6-phosphate phosphatase   \n",
              "GAPD               Glyceraldehyde-3-phosphate dehydrogenase   \n",
              "GLBRAN2   1,4-alpha-glucan branching enzyme (glycogen ->...   \n",
              "GLCP                                 Glycogen phosphorylase   \n",
              "GLCP2                                Glycogen phosphorylase   \n",
              "GLCS1                            Glycogen synthase (ADPGlc)   \n",
              "GLDBRAN2  Glycogen debranching enzyme (bglycogen -> glyc...   \n",
              "GLGC                Glucose-1-phosphate adenylyltransferase   \n",
              "HEX1                             Hexokinase (D-glucose:ATP)   \n",
              "PDH                                  Pyruvate dehydrogenase   \n",
              "PFK                                     Phosphofructokinase   \n",
              "PGI                           Glucose-6-phosphate isomerase   \n",
              "PGK                                 Phosphoglycerate kinase   \n",
              "PGM                                 Phosphoglycerate mutase   \n",
              "PPS                            Phosphoenolpyruvate synthase   \n",
              "PYK                                         Pyruvate kinase   \n",
              "TPI                              Triose-phosphate isomerase   \n",
              "\n",
              "                                          Reaction_Formulas  \n",
              "ENO                                 2pg_c <=> h2o_c + pep_c  \n",
              "F6PA                                f6p_c <=> dha_c + g3p_c  \n",
              "FBA                                fdp_c <=> dhap_c + g3p_c  \n",
              "FBP                          fdp_c + h2o_c --> f6p_c + pi_c  \n",
              "G1PPpp                    g1p_p + h2o_p --> glc__D_p + pi_p  \n",
              "G6PP                      g6p_c + h2o_c --> glc__D_c + pi_c  \n",
              "GAPD        g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c  \n",
              "GLBRAN2                          glycogen_c --> bglycogen_c  \n",
              "GLCP                            glycogen_c + pi_c --> g1p_c  \n",
              "GLCP2                          bglycogen_c + pi_c --> g1p_c  \n",
              "GLCS1                 adpglc_c --> adp_c + glycogen_c + h_c  \n",
              "GLDBRAN2                         bglycogen_c --> glycogen_c  \n",
              "GLGC               atp_c + g1p_c + h_c --> adpglc_c + ppi_c  \n",
              "HEX1               atp_c + glc__D_c --> adp_c + g6p_c + h_c  \n",
              "PDH       coa_c + nad_c + pyr_c --> accoa_c + co2_c + na...  \n",
              "PFK                   atp_c + f6p_c --> adp_c + fdp_c + h_c  \n",
              "PGI                                         g6p_c <=> f6p_c  \n",
              "PGK                       3pg_c + atp_c <=> 13dpg_c + adp_c  \n",
              "PGM                                         2pg_c <=> 3pg_c  \n",
              "PPS       atp_c + h2o_c + pyr_c --> amp_c + 2.0 h_c + pe...  \n",
              "PYK                   adp_c + h_c + pep_c --> atp_c + pyr_c  \n",
              "TPI                                        dhap_c <=> g3p_c  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-29a92aaa-9583-4745-ab54-ea8e97f6be2b\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Reaction_Names</th>\n",
              "      <th>Reaction_Formulas</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENO</th>\n",
              "      <td>Enolase</td>\n",
              "      <td>2pg_c &lt;=&gt; h2o_c + pep_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F6PA</th>\n",
              "      <td>Fructose 6-phosphate aldolase</td>\n",
              "      <td>f6p_c &lt;=&gt; dha_c + g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>FBA</th>\n",
              "      <td>Fructose-bisphosphate aldolase</td>\n",
              "      <td>fdp_c &lt;=&gt; dhap_c + g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>FBP</th>\n",
              "      <td>Fructose-bisphosphatase</td>\n",
              "      <td>fdp_c + h2o_c --&gt; f6p_c + pi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G1PPpp</th>\n",
              "      <td>Glucose-1-phosphatase</td>\n",
              "      <td>g1p_p + h2o_p --&gt; glc__D_p + pi_p</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>G6PP</th>\n",
              "      <td>Glucose-6-phosphate phosphatase</td>\n",
              "      <td>g6p_c + h2o_c --&gt; glc__D_c + pi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GAPD</th>\n",
              "      <td>Glyceraldehyde-3-phosphate dehydrogenase</td>\n",
              "      <td>g3p_c + nad_c + pi_c &lt;=&gt; 13dpg_c + h_c + nadh_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GLBRAN2</th>\n",
              "      <td>1,4-alpha-glucan branching enzyme (glycogen -&gt;...</td>\n",
              "      <td>glycogen_c --&gt; bglycogen_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GLCP</th>\n",
              "      <td>Glycogen phosphorylase</td>\n",
              "      <td>glycogen_c + pi_c --&gt; g1p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GLCP2</th>\n",
              "      <td>Glycogen phosphorylase</td>\n",
              "      <td>bglycogen_c + pi_c --&gt; g1p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GLCS1</th>\n",
              "      <td>Glycogen synthase (ADPGlc)</td>\n",
              "      <td>adpglc_c --&gt; adp_c + glycogen_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GLDBRAN2</th>\n",
              "      <td>Glycogen debranching enzyme (bglycogen -&gt; glyc...</td>\n",
              "      <td>bglycogen_c --&gt; glycogen_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GLGC</th>\n",
              "      <td>Glucose-1-phosphate adenylyltransferase</td>\n",
              "      <td>atp_c + g1p_c + h_c --&gt; adpglc_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HEX1</th>\n",
              "      <td>Hexokinase (D-glucose:ATP)</td>\n",
              "      <td>atp_c + glc__D_c --&gt; adp_c + g6p_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PDH</th>\n",
              "      <td>Pyruvate dehydrogenase</td>\n",
              "      <td>coa_c + nad_c + pyr_c --&gt; accoa_c + co2_c + na...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PFK</th>\n",
              "      <td>Phosphofructokinase</td>\n",
              "      <td>atp_c + f6p_c --&gt; adp_c + fdp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGI</th>\n",
              "      <td>Glucose-6-phosphate isomerase</td>\n",
              "      <td>g6p_c &lt;=&gt; f6p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGK</th>\n",
              "      <td>Phosphoglycerate kinase</td>\n",
              "      <td>3pg_c + atp_c &lt;=&gt; 13dpg_c + adp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGM</th>\n",
              "      <td>Phosphoglycerate mutase</td>\n",
              "      <td>2pg_c &lt;=&gt; 3pg_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PPS</th>\n",
              "      <td>Phosphoenolpyruvate synthase</td>\n",
              "      <td>atp_c + h2o_c + pyr_c --&gt; amp_c + 2.0 h_c + pe...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PYK</th>\n",
              "      <td>Pyruvate kinase</td>\n",
              "      <td>adp_c + h_c + pep_c --&gt; atp_c + pyr_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TPI</th>\n",
              "      <td>Triose-phosphate isomerase</td>\n",
              "      <td>dhap_c &lt;=&gt; g3p_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-29a92aaa-9583-4745-ab54-ea8e97f6be2b')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-29a92aaa-9583-4745-ab54-ea8e97f6be2b button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-29a92aaa-9583-4745-ab54-ea8e97f6be2b');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 330
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Because we are interested in building the lycopene pathway, the first thing we need to do is extract the reactions required for the pathway."
      ],
      "metadata": {
        "id": "XpJe9RV-gM86"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# take out all reactions from the Glycolysis/Gluconeogenesis\n",
        "our_glycolysisReactions = [r for r in glycolysisReactions if r.id not in ['F6PA', 'FBP','G1PPpp','GLBRAN2','GLCP','GLCP2','GLCS1','GLDBRAN2','GLGC','PDH','PPS','G6PP']]\n",
        "\n",
        "# Create a table for the reactions\n",
        "reactionNames = [r.name for r in our_glycolysisReactions]\n",
        "reactionFormulas = [r.build_reaction_string() for r in our_glycolysisReactions]\n",
        "tableData = {'Reaction_Names': reactionNames, 'Reaction_Formulas': reactionFormulas}\n",
        "tableIndex = [r.id for r in our_glycolysisReactions]\n",
        "T = pd.DataFrame(tableData, index=tableIndex)\n",
        "T"
      ],
      "metadata": {
        "id": "b1VRNc3ZODeE",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 363
        },
        "outputId": "7243e5a0-5de1-4754-acb7-362c7b66d492"
      },
      "execution_count": 331,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                Reaction_Names  \\\n",
              "ENO                                    Enolase   \n",
              "FBA             Fructose-bisphosphate aldolase   \n",
              "GAPD  Glyceraldehyde-3-phosphate dehydrogenase   \n",
              "HEX1                Hexokinase (D-glucose:ATP)   \n",
              "PFK                        Phosphofructokinase   \n",
              "PGI              Glucose-6-phosphate isomerase   \n",
              "PGK                    Phosphoglycerate kinase   \n",
              "PGM                    Phosphoglycerate mutase   \n",
              "PYK                            Pyruvate kinase   \n",
              "TPI                 Triose-phosphate isomerase   \n",
              "\n",
              "                                    Reaction_Formulas  \n",
              "ENO                           2pg_c <=> h2o_c + pep_c  \n",
              "FBA                          fdp_c <=> dhap_c + g3p_c  \n",
              "GAPD  g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c  \n",
              "HEX1         atp_c + glc__D_c --> adp_c + g6p_c + h_c  \n",
              "PFK             atp_c + f6p_c --> adp_c + fdp_c + h_c  \n",
              "PGI                                   g6p_c <=> f6p_c  \n",
              "PGK                 3pg_c + atp_c <=> 13dpg_c + adp_c  \n",
              "PGM                                   2pg_c <=> 3pg_c  \n",
              "PYK             adp_c + h_c + pep_c --> atp_c + pyr_c  \n",
              "TPI                                  dhap_c <=> g3p_c  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3c9d7c88-421a-4c61-aa2d-4a1880e1ea19\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Reaction_Names</th>\n",
              "      <th>Reaction_Formulas</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENO</th>\n",
              "      <td>Enolase</td>\n",
              "      <td>2pg_c &lt;=&gt; h2o_c + pep_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>FBA</th>\n",
              "      <td>Fructose-bisphosphate aldolase</td>\n",
              "      <td>fdp_c &lt;=&gt; dhap_c + g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GAPD</th>\n",
              "      <td>Glyceraldehyde-3-phosphate dehydrogenase</td>\n",
              "      <td>g3p_c + nad_c + pi_c &lt;=&gt; 13dpg_c + h_c + nadh_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HEX1</th>\n",
              "      <td>Hexokinase (D-glucose:ATP)</td>\n",
              "      <td>atp_c + glc__D_c --&gt; adp_c + g6p_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PFK</th>\n",
              "      <td>Phosphofructokinase</td>\n",
              "      <td>atp_c + f6p_c --&gt; adp_c + fdp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGI</th>\n",
              "      <td>Glucose-6-phosphate isomerase</td>\n",
              "      <td>g6p_c &lt;=&gt; f6p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGK</th>\n",
              "      <td>Phosphoglycerate kinase</td>\n",
              "      <td>3pg_c + atp_c &lt;=&gt; 13dpg_c + adp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGM</th>\n",
              "      <td>Phosphoglycerate mutase</td>\n",
              "      <td>2pg_c &lt;=&gt; 3pg_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PYK</th>\n",
              "      <td>Pyruvate kinase</td>\n",
              "      <td>adp_c + h_c + pep_c --&gt; atp_c + pyr_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TPI</th>\n",
              "      <td>Triose-phosphate isomerase</td>\n",
              "      <td>dhap_c &lt;=&gt; g3p_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3c9d7c88-421a-4c61-aa2d-4a1880e1ea19')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3c9d7c88-421a-4c61-aa2d-4a1880e1ea19 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3c9d7c88-421a-4c61-aa2d-4a1880e1ea19');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 331
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Now, we require the methylerythritol phosphate pathway I for our analysis. Since we only have a partial set of reactions for this pathway and there is no subsystem assigned to it, we will implement it manually."
      ],
      "metadata": {
        "id": "-AG9vLwShKcW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import difflib\n",
        "\n",
        "# this method is useful for extract our reactions. \n",
        "def find_reaction(enzyme: str):\n",
        "  highest_score = 0\n",
        "  reaction = \"\"\n",
        "  for rct in ecoli_model.reactions:\n",
        "    matcher = difflib.SequenceMatcher(None, enzyme, rct.name)\n",
        "    if matcher.quick_ratio() >= highest_score:\n",
        "      highest_score = matcher.quick_ratio()\n",
        "      reaction = rct\n",
        "  return reaction,highest_score\n",
        "# method is not that good but it is very helpful when you have 9 or above score of Levenshtein distance. \n",
        "# my last 3 reactions was not good so I did look up the other names from KEGG and put it here. "
      ],
      "metadata": {
        "id": "jnL8UW_PhLU4"
      },
      "execution_count": 332,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# extract the partial methylerythritol phosphate pathway\n",
        "partial_methylerythritol_phosphate = []\n",
        "scores = []\n",
        "\n",
        "all_enzymes = ['1-deoxy-D-xylulose-5-phosphate synthase', '1-deoxy-D-xylulose-5-phosphate reductoisomerase', '2-C-methyl-D-erythritol 4-phosphate cytidylyltransferase',\n",
        "               '4-(cytidine 5-diphospho)-2-C-methyl-D-erythritol kinase','2-C-methyl-D-erythritol 2,4-cyclodiphosphate synthase', '2C-methyl-D-erythritol 2,4 cyclodiphosphate dehydratase', \n",
        "               '1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate reductase (ipdp)', '1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate reductase (dmpp)', 'isopenthenyl-diphosphate isomerase'\n",
        "               , 'Dimethylallyltranstransferase', 'Geranyltranstransferase']\n",
        "for enz in all_enzymes:\n",
        "  rct, score = find_reaction(enz)\n",
        "  partial_methylerythritol_phosphate.append(rct)\n",
        "  scores.append(score)\n",
        "\n",
        "reactionNames = [r.name for r in partial_methylerythritol_phosphate]\n",
        "reactionFormulas = [r.build_reaction_string() for r in partial_methylerythritol_phosphate]\n",
        "tableData = {'Reaction_Names': reactionNames, 'Reaction_Formulas': reactionFormulas}\n",
        "tableIndex = [r.id for r in partial_methylerythritol_phosphate]\n",
        "T = pd.DataFrame(tableData, index=tableIndex)\n",
        "T"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 394
        },
        "id": "_O_UIRzejW1U",
        "outputId": "91e63990-7544-4f4b-ce31-6733980afa12"
      },
      "execution_count": 333,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                             Reaction_Names  \\\n",
              "DXPS                1-deoxy-D-xylulose 5-phosphate synthase   \n",
              "DXPRIi                  1-deoxy-D-xylulose reductoisomerase   \n",
              "MEPCT     2-C-methyl-D-erythritol 4-phosphate cytidylylt...   \n",
              "CDPMEK    4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...   \n",
              "MECDPS    2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...   \n",
              "MECDPDH5  2C-methyl-D-erythritol 2,4 cyclodiphosphate de...   \n",
              "IPDPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...   \n",
              "DMPPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...   \n",
              "IPDDI                   Isopentenyl-diphosphate D-isomerase   \n",
              "DMATT                         Dimethylallyltranstransferase   \n",
              "GRTT                                Geranyltranstransferase   \n",
              "\n",
              "                                          Reaction_Formulas  \n",
              "DXPS               g3p_c + h_c + pyr_c --> co2_c + dxyl5p_c  \n",
              "DXPRIi        dxyl5p_c + h_c + nadph_c --> 2me4p_c + nadp_c  \n",
              "MEPCT             2me4p_c + ctp_c + h_c --> 4c2me_c + ppi_c  \n",
              "CDPMEK          4c2me_c + atp_c --> 2p4c2me_c + adp_c + h_c  \n",
              "MECDPS                       2p4c2me_c --> 2mecdp_c + cmp_c  \n",
              "MECDPDH5  2mecdp_c + 2.0 flxr_c + h_c --> 2.0 flxso_c + ...  \n",
              "IPDPS     h2mb4p_c + h_c + nadh_c --> h2o_c + ipdp_c + n...  \n",
              "DMPPS     h2mb4p_c + h_c + nadh_c --> dmpp_c + h2o_c + n...  \n",
              "IPDDI                                     ipdp_c <=> dmpp_c  \n",
              "DMATT                    dmpp_c + ipdp_c --> grdp_c + ppi_c  \n",
              "GRTT                     grdp_c + ipdp_c --> frdp_c + ppi_c  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-bbbd0929-4958-4a45-958d-ba82c0b2554a\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Reaction_Names</th>\n",
              "      <th>Reaction_Formulas</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>DXPS</th>\n",
              "      <td>1-deoxy-D-xylulose 5-phosphate synthase</td>\n",
              "      <td>g3p_c + h_c + pyr_c --&gt; co2_c + dxyl5p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DXPRIi</th>\n",
              "      <td>1-deoxy-D-xylulose reductoisomerase</td>\n",
              "      <td>dxyl5p_c + h_c + nadph_c --&gt; 2me4p_c + nadp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MEPCT</th>\n",
              "      <td>2-C-methyl-D-erythritol 4-phosphate cytidylylt...</td>\n",
              "      <td>2me4p_c + ctp_c + h_c --&gt; 4c2me_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CDPMEK</th>\n",
              "      <td>4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...</td>\n",
              "      <td>4c2me_c + atp_c --&gt; 2p4c2me_c + adp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPS</th>\n",
              "      <td>2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...</td>\n",
              "      <td>2p4c2me_c --&gt; 2mecdp_c + cmp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPDH5</th>\n",
              "      <td>2C-methyl-D-erythritol 2,4 cyclodiphosphate de...</td>\n",
              "      <td>2mecdp_c + 2.0 flxr_c + h_c --&gt; 2.0 flxso_c + ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; h2o_c + ipdp_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMPPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; dmpp_c + h2o_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDDI</th>\n",
              "      <td>Isopentenyl-diphosphate D-isomerase</td>\n",
              "      <td>ipdp_c &lt;=&gt; dmpp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMATT</th>\n",
              "      <td>Dimethylallyltranstransferase</td>\n",
              "      <td>dmpp_c + ipdp_c --&gt; grdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GRTT</th>\n",
              "      <td>Geranyltranstransferase</td>\n",
              "      <td>grdp_c + ipdp_c --&gt; frdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-bbbd0929-4958-4a45-958d-ba82c0b2554a')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-bbbd0929-4958-4a45-958d-ba82c0b2554a button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-bbbd0929-4958-4a45-958d-ba82c0b2554a');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 333
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Our lycopene pathway is incomplete, as there are three additional steps that have yet to be included in our model. We need to incorporate these steps in order to analyze the pathway effectively.\n",
        "\n",
        "1) (2E,6E)-farnesyl diphosphate + 3-methylbut-3-en-1-yl diphosphate -> geranylgeranyl diphosphate + diphosphate <br>\n",
        "2) 2 geranylgeranyl diphosphate -> 15-cis-phyteone + 2 diphosphate <br>\n",
        "3) 15-cis-phyteone + 4 FAD -> all-trans-lycopene + 4 FADH2<br>\n",
        "\n",
        "For the final reaction, the electron carrier could be either FAD or NAD+ as a cofactor, although FAD has a lower probability of serving as the electron carrier compared to NAD+"
      ],
      "metadata": {
        "id": "3ipZq_P99nSC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# example for showing instance variables of Metabolites. \n",
        "ecoli_model.metabolites.frdp_c\n",
        "ecoli_model.metabolites.get_by_id('ppi_c')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 177
        },
        "id": "K8vzxMHkNxTD",
        "outputId": "37a68061-7827-4923-cc4c-5b1f28d6f3a1"
      },
      "execution_count": 334,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Metabolite ppi_c at 0x7f5a103c87f0>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Metabolite identifier</strong></td><td>ppi_c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>Diphosphate</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a103c87f0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Formula</strong></td><td>HO7P2</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Compartment</strong></td><td>c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>In 145 reaction(s)</strong></td><td>\n",
              "                    NTPP4, PROTRS, NTPP5, 2AGPEAT160, PHETRS, GALUi, GLGC, THRTRS, BMOGDS2, DASYN180, GLUPRT, 2AGPEAT181, DHBS, GDPTPDP, GMPS2, 2AGPGAT180, DASYN120, ADNCYC, GLUTRS, DASYN181, DASYN160, FACOAL181t2pp,...\n",
              "                    </td>\n",
              "            </tr>\n",
              "        </table>"
            ]
          },
          "metadata": {},
          "execution_count": 334
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from cobra.core import Metabolite\n",
        "\n",
        "# we have to create metabolites first\n",
        "# Creat three metabolites.\n",
        "geranyl_diphosphate = Metabolite('ggdp_c', name = 'Geranylgeranyl diphosphate',\n",
        "                                 compartment='c', formula= 'C20H36O7P2', charge = -3)\n",
        "cis_phyteone = Metabolite('15cp_c', name = '15-cis-phyteone', compartment = 'c',\n",
        "                          formula= 'C40H64', charge=0)\n",
        "lycopene = Metabolite('lcp_c', name= 'All-trans-lycopene', compartment= 'c', \n",
        "                      formula='C40H56', charge = 0)\n",
        "\n",
        "# Add the metabolites to the model\n",
        "ecoli_model.add_metabolites([geranyl_diphosphate,cis_phyteone, lycopene])\n",
        "\n",
        "# check your work\n",
        "# after you add reactions, you will see them\n",
        "ecoli_model.metabolites.ggdp_c"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "bSPC_eql9m1s",
        "outputId": "03c034bd-e591-4600-876f-6d1d19aa87bc"
      },
      "execution_count": 335,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Metabolite ggdp_c at 0x7f5a0eb58730>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Metabolite identifier</strong></td><td>ggdp_c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>Geranylgeranyl diphosphate</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0eb58730</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Formula</strong></td><td>C20H36O7P2</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Compartment</strong></td><td>c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>In 0 reaction(s)</strong></td><td>\n",
              "                    \n",
              "                    </td>\n",
              "            </tr>\n",
              "        </table>"
            ]
          },
          "metadata": {},
          "execution_count": 335
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# check your WindowingRankType.\n",
        "ecoli_model.metabolites.get_by_id('15cp_c') # It is safer to use an ID with a number at the front."
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "mfcdLxr7oupg",
        "outputId": "51c761b4-8024-4a19-d41b-95fa60d3d1dc"
      },
      "execution_count": 336,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Metabolite 15cp_c at 0x7f5a1b502a60>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Metabolite identifier</strong></td><td>15cp_c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>15-cis-phyteone</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a1b502a60</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Formula</strong></td><td>C40H64</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Compartment</strong></td><td>c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>In 0 reaction(s)</strong></td><td>\n",
              "                    \n",
              "                    </td>\n",
              "            </tr>\n",
              "        </table>"
            ]
          },
          "metadata": {},
          "execution_count": 336
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# check your work\n",
        "ecoli_model.metabolites.lcp_c"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "V2DCZOkwpkPH",
        "outputId": "1e61e4a9-db8c-4d1f-9781-09d1444446d5"
      },
      "execution_count": 337,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Metabolite lcp_c at 0x7f5a0cfc6c70>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Metabolite identifier</strong></td><td>lcp_c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>All-trans-lycopene</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0cfc6c70</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Formula</strong></td><td>C40H56</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Compartment</strong></td><td>c</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>In 0 reaction(s)</strong></td><td>\n",
              "                    \n",
              "                    </td>\n",
              "            </tr>\n",
              "        </table>"
            ]
          },
          "metadata": {},
          "execution_count": 337
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We just added the required metabolites to our model. Now, the reactions"
      ],
      "metadata": {
        "id": "mncTVfy0WWwe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# example for showing instance variables of Reaction. \n",
        "ecoli_model.genes.b0421\n",
        "ecoli_model.reactions.GRTT\n",
        "ecoli_model.reactions.get_by_id('GRTT')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "zJpUabTGWrjP",
        "outputId": "3a2c0beb-faeb-4fdf-d12b-533529f551b3"
      },
      "execution_count": 338,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction GRTT at 0x7f5a0c5d96d0>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>GRTT</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>Geranyltranstransferase</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0c5d96d0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>grdp_c + ipdp_c --> frdp_c + ppi_c</p>\n",
              "                    <p style='text-align:right'>Geranyl diphosphate + Isopentenyl diphosphate --> Farnesyl diphosphate + Diphosphate</p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td>b0421</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 338
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from cobra.core import Reaction\n",
        "from cobra.core import Gene\n",
        "\n",
        "# just for making things simple\n",
        "farnesyl_diphosphate = ecoli_model.metabolites.frdp_c\n",
        "isopentenyl_diphosphate = ecoli_model.metabolites.ipdp_c #  Isopentenyl diphosphate is same as 3-methylbut-3-en-1-yl diphosphate\n",
        "diphosphate = ecoli_model.metabolites.ppi_c\n",
        "fad = ecoli_model.metabolites.fad_c\n",
        "fadh2 = ecoli_model.metabolites.fadh2_c\n",
        "\n",
        "# (2E,6E)-farnesyl diphosphate + 3-methylbut-3-en-1-yl diphosphate \n",
        "# -> geranylgeranyl diphosphate + diphosphate\n",
        "reaction_1 = Reaction('GDS', name= 'geranylgeranyl diphosphate synthase', \n",
        "                      subsystem= 'Lycopene pathway', lower_bound= 0\n",
        "                      , upper_bound=1000) # You can set your bounds later if you want to.\n",
        "reaction_1.add_metabolites({farnesyl_diphosphate: -1, isopentenyl_diphosphate:-1, \n",
        "                            geranyl_diphosphate: 1, diphosphate: 1})\n",
        "reaction_1.gene_reaction_rule = 'b5001' # Since it is new gene. our model genes ends around b4500\n",
        "\n",
        "# 2 geranylgeranyl diphosphate -> 15-cis-phyteone + 2 diphosphate\n",
        "reaction_2 = Reaction('CPS',name = '15-cis-phytoene synthase', subsystem ='Lycopene pathway',\n",
        "                      lower_bound=0, upper_bound= 1000)\n",
        "reaction_2.add_metabolites({geranyl_diphosphate: -2, cis_phyteone: 1, diphosphate: 2})\n",
        "reaction_2.gene_reaction_rule = 'b5002'\n",
        "\n",
        "# 15-cis-phyteone + 4 FAD -> all-trans-lycopene + 4 FADH2\n",
        "reaction_3 = Reaction('PDS', name= 'phytoene desaturase', subsystem= 'Lycopene pathway',\n",
        "                      lower_bound= 0, upper_bound= 1000)\n",
        "reaction_3.add_metabolites({cis_phyteone: -1, fad: -4, lycopene: 1, fadh2: 4})\n",
        "reaction_3.gene_reaction_rule = 'b5003'\n",
        "\n",
        "ecoli_model.add_reactions([reaction_1, reaction_2, reaction_3])\n",
        "ecoli_model.reactions.GDS"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "4UPJFe9IWHGF",
        "outputId": "7e3cee98-96a7-49c9-f70c-b1e0bbc7ab70"
      },
      "execution_count": 339,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction GDS at 0x7f5a0b9d0a00>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>GDS</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>geranylgeranyl diphosphate synthase</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0b9d0a00</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>frdp_c + ipdp_c --> ggdp_c + ppi_c</p>\n",
              "                    <p style='text-align:right'>Farnesyl diphosphate + Isopentenyl diphosphate --> Geranylgeranyl diphosphate + Diphosphate</p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td>b5001</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 339
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# check your work\n",
        "ecoli_model.reactions.CPS"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "JXIGzv9XncyD",
        "outputId": "917dc15a-25a0-4d91-ec57-e97ecc93e2c9"
      },
      "execution_count": 340,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction CPS at 0x7f5a0b9d0430>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>CPS</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>15-cis-phytoene synthase</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0b9d0430</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>2 ggdp_c --> 15cp_c + 2 ppi_c</p>\n",
              "                    <p style='text-align:right'>2 Geranylgeranyl diphosphate --> 15-cis-phyteone + 2 Diphosphate</p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td>b5002</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 340
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# check your work\n",
        "ecoli_model.reactions.PDS"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "o-5yqG8tocjh",
        "outputId": "4af8d723-4c98-4b40-a034-2a0835688f28"
      },
      "execution_count": 341,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction PDS at 0x7f5a0b9d06d0>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>PDS</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>phytoene desaturase</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0b9d06d0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>15cp_c + 4 fad_c --> 4 fadh2_c + lcp_c</p>\n",
              "                    <p style='text-align:right'>15-cis-phyteone + 4 Flavin adenine dinucleotide oxidized --> 4 Flavin adenine dinucleotide reduced + All-trans-lycopene</p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td>b5003</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 341
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Since our purpose is to analyze the yield of metabolites, setting up the genes is unnecessary. An example where using genes would be applicable is when knocking out a gene.<br>\n",
        "\n",
        "We have all the metabolites and reactions we need, so the next step will be to add the subsystem for the lycopene pathway."
      ],
      "metadata": {
        "id": "VpdwWTL9p9Ie"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# collect the all reactions to build the pathway \n",
        "\n",
        "reactions_lycophene_pathway = our_glycolysisReactions + partial_methylerythritol_phosphate\n",
        "reactions_lycophene_pathway.append(reaction_1)\n",
        "reactions_lycophene_pathway.append(reaction_2)\n",
        "reactions_lycophene_pathway.append(reaction_3)\n",
        "reactions_lycophene_pathway\n",
        "reactionNames = [r.name for r in reactions_lycophene_pathway]\n",
        "reactionFormulas = [r.build_reaction_string() for r in reactions_lycophene_pathway]\n",
        "tableData = {'Reaction_Names': reactionNames, 'Reaction_Formulas': reactionFormulas}\n",
        "tableIndex = [r.id for r in reactions_lycophene_pathway]\n",
        "T = pd.DataFrame(tableData, index=tableIndex)\n",
        "T"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 802
        },
        "id": "RbP4ntGQrWXs",
        "outputId": "2232cf07-83d9-49d6-ed24-cb622342faf8"
      },
      "execution_count": 342,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                             Reaction_Names  \\\n",
              "ENO                                                 Enolase   \n",
              "FBA                          Fructose-bisphosphate aldolase   \n",
              "GAPD               Glyceraldehyde-3-phosphate dehydrogenase   \n",
              "HEX1                             Hexokinase (D-glucose:ATP)   \n",
              "PFK                                     Phosphofructokinase   \n",
              "PGI                           Glucose-6-phosphate isomerase   \n",
              "PGK                                 Phosphoglycerate kinase   \n",
              "PGM                                 Phosphoglycerate mutase   \n",
              "PYK                                         Pyruvate kinase   \n",
              "TPI                              Triose-phosphate isomerase   \n",
              "DXPS                1-deoxy-D-xylulose 5-phosphate synthase   \n",
              "DXPRIi                  1-deoxy-D-xylulose reductoisomerase   \n",
              "MEPCT     2-C-methyl-D-erythritol 4-phosphate cytidylylt...   \n",
              "CDPMEK    4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...   \n",
              "MECDPS    2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...   \n",
              "MECDPDH5  2C-methyl-D-erythritol 2,4 cyclodiphosphate de...   \n",
              "IPDPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...   \n",
              "DMPPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...   \n",
              "IPDDI                   Isopentenyl-diphosphate D-isomerase   \n",
              "DMATT                         Dimethylallyltranstransferase   \n",
              "GRTT                                Geranyltranstransferase   \n",
              "GDS                     geranylgeranyl diphosphate synthase   \n",
              "CPS                                15-cis-phytoene synthase   \n",
              "PDS                                     phytoene desaturase   \n",
              "\n",
              "                                          Reaction_Formulas  \n",
              "ENO                                 2pg_c <=> h2o_c + pep_c  \n",
              "FBA                                fdp_c <=> dhap_c + g3p_c  \n",
              "GAPD        g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c  \n",
              "HEX1               atp_c + glc__D_c --> adp_c + g6p_c + h_c  \n",
              "PFK                   atp_c + f6p_c --> adp_c + fdp_c + h_c  \n",
              "PGI                                         g6p_c <=> f6p_c  \n",
              "PGK                       3pg_c + atp_c <=> 13dpg_c + adp_c  \n",
              "PGM                                         2pg_c <=> 3pg_c  \n",
              "PYK                   adp_c + h_c + pep_c --> atp_c + pyr_c  \n",
              "TPI                                        dhap_c <=> g3p_c  \n",
              "DXPS               g3p_c + h_c + pyr_c --> co2_c + dxyl5p_c  \n",
              "DXPRIi        dxyl5p_c + h_c + nadph_c --> 2me4p_c + nadp_c  \n",
              "MEPCT             2me4p_c + ctp_c + h_c --> 4c2me_c + ppi_c  \n",
              "CDPMEK          4c2me_c + atp_c --> 2p4c2me_c + adp_c + h_c  \n",
              "MECDPS                       2p4c2me_c --> 2mecdp_c + cmp_c  \n",
              "MECDPDH5  2mecdp_c + 2.0 flxr_c + h_c --> 2.0 flxso_c + ...  \n",
              "IPDPS     h2mb4p_c + h_c + nadh_c --> h2o_c + ipdp_c + n...  \n",
              "DMPPS     h2mb4p_c + h_c + nadh_c --> dmpp_c + h2o_c + n...  \n",
              "IPDDI                                     ipdp_c <=> dmpp_c  \n",
              "DMATT                    dmpp_c + ipdp_c --> grdp_c + ppi_c  \n",
              "GRTT                     grdp_c + ipdp_c --> frdp_c + ppi_c  \n",
              "GDS                      frdp_c + ipdp_c --> ggdp_c + ppi_c  \n",
              "CPS                           2 ggdp_c --> 15cp_c + 2 ppi_c  \n",
              "PDS                  15cp_c + 4 fad_c --> 4 fadh2_c + lcp_c  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-6e180c82-04a0-41d9-9e5f-73610ea80cba\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Reaction_Names</th>\n",
              "      <th>Reaction_Formulas</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENO</th>\n",
              "      <td>Enolase</td>\n",
              "      <td>2pg_c &lt;=&gt; h2o_c + pep_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>FBA</th>\n",
              "      <td>Fructose-bisphosphate aldolase</td>\n",
              "      <td>fdp_c &lt;=&gt; dhap_c + g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GAPD</th>\n",
              "      <td>Glyceraldehyde-3-phosphate dehydrogenase</td>\n",
              "      <td>g3p_c + nad_c + pi_c &lt;=&gt; 13dpg_c + h_c + nadh_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HEX1</th>\n",
              "      <td>Hexokinase (D-glucose:ATP)</td>\n",
              "      <td>atp_c + glc__D_c --&gt; adp_c + g6p_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PFK</th>\n",
              "      <td>Phosphofructokinase</td>\n",
              "      <td>atp_c + f6p_c --&gt; adp_c + fdp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGI</th>\n",
              "      <td>Glucose-6-phosphate isomerase</td>\n",
              "      <td>g6p_c &lt;=&gt; f6p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGK</th>\n",
              "      <td>Phosphoglycerate kinase</td>\n",
              "      <td>3pg_c + atp_c &lt;=&gt; 13dpg_c + adp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGM</th>\n",
              "      <td>Phosphoglycerate mutase</td>\n",
              "      <td>2pg_c &lt;=&gt; 3pg_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PYK</th>\n",
              "      <td>Pyruvate kinase</td>\n",
              "      <td>adp_c + h_c + pep_c --&gt; atp_c + pyr_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TPI</th>\n",
              "      <td>Triose-phosphate isomerase</td>\n",
              "      <td>dhap_c &lt;=&gt; g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DXPS</th>\n",
              "      <td>1-deoxy-D-xylulose 5-phosphate synthase</td>\n",
              "      <td>g3p_c + h_c + pyr_c --&gt; co2_c + dxyl5p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DXPRIi</th>\n",
              "      <td>1-deoxy-D-xylulose reductoisomerase</td>\n",
              "      <td>dxyl5p_c + h_c + nadph_c --&gt; 2me4p_c + nadp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MEPCT</th>\n",
              "      <td>2-C-methyl-D-erythritol 4-phosphate cytidylylt...</td>\n",
              "      <td>2me4p_c + ctp_c + h_c --&gt; 4c2me_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CDPMEK</th>\n",
              "      <td>4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...</td>\n",
              "      <td>4c2me_c + atp_c --&gt; 2p4c2me_c + adp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPS</th>\n",
              "      <td>2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...</td>\n",
              "      <td>2p4c2me_c --&gt; 2mecdp_c + cmp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPDH5</th>\n",
              "      <td>2C-methyl-D-erythritol 2,4 cyclodiphosphate de...</td>\n",
              "      <td>2mecdp_c + 2.0 flxr_c + h_c --&gt; 2.0 flxso_c + ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; h2o_c + ipdp_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMPPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; dmpp_c + h2o_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDDI</th>\n",
              "      <td>Isopentenyl-diphosphate D-isomerase</td>\n",
              "      <td>ipdp_c &lt;=&gt; dmpp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMATT</th>\n",
              "      <td>Dimethylallyltranstransferase</td>\n",
              "      <td>dmpp_c + ipdp_c --&gt; grdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GRTT</th>\n",
              "      <td>Geranyltranstransferase</td>\n",
              "      <td>grdp_c + ipdp_c --&gt; frdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GDS</th>\n",
              "      <td>geranylgeranyl diphosphate synthase</td>\n",
              "      <td>frdp_c + ipdp_c --&gt; ggdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CPS</th>\n",
              "      <td>15-cis-phytoene synthase</td>\n",
              "      <td>2 ggdp_c --&gt; 15cp_c + 2 ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PDS</th>\n",
              "      <td>phytoene desaturase</td>\n",
              "      <td>15cp_c + 4 fad_c --&gt; 4 fadh2_c + lcp_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-6e180c82-04a0-41d9-9e5f-73610ea80cba')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-6e180c82-04a0-41d9-9e5f-73610ea80cba button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-6e180c82-04a0-41d9-9e5f-73610ea80cba');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 342
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We need three more reactions to add later. For now, they are good enough."
      ],
      "metadata": {
        "id": "jiGYKRSSyLoB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "for rct in reactions_lycophene_pathway:\n",
        "  subsystem = rct.subsystem\n",
        "   # there is no Lycopene pathway but it is necessary for rerunning the code\n",
        "  if subsystem == 'Lycopene pathway':\n",
        "    continue\n",
        "  elif type(subsystem) is list and 'Lycopene pathway' not in subsystem:\n",
        "    rct.subsystem = rct.subsystem.append('Lycopene pathway')\n",
        "  elif subsystem == '':\n",
        "    rct.subsystem = 'Lycopene pathway'\n",
        "  elif type(subsystem) == str:\n",
        "    sub_list = []\n",
        "    sub_list.append(subsystem)\n",
        "    sub_list.append('Lycopene pathway')\n",
        "    rct.subsystem = sub_list\n",
        "\n",
        "reactionNames = [r.name for r in reactions_lycophene_pathway]\n",
        "reactionFormulas = [r.build_reaction_string() for r in reactions_lycophene_pathway]\n",
        "subsystemNames = [r.subsystem for r in reactions_lycophene_pathway]\n",
        "tableData = {'Reaction_Names': reactionNames, 'Reaction_subsystem' : subsystemNames, 'Reaction_Formulas': reactionFormulas}\n",
        "tableIndex = [r.id for r in reactions_lycophene_pathway]\n",
        "T = pd.DataFrame(tableData, index=tableIndex)\n",
        "T\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 802
        },
        "id": "UNaeyHuLydPL",
        "outputId": "5de9ce1c-f0ae-472f-e55d-8acb06da6d8c"
      },
      "execution_count": 343,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                             Reaction_Names  \\\n",
              "ENO                                                 Enolase   \n",
              "FBA                          Fructose-bisphosphate aldolase   \n",
              "GAPD               Glyceraldehyde-3-phosphate dehydrogenase   \n",
              "HEX1                             Hexokinase (D-glucose:ATP)   \n",
              "PFK                                     Phosphofructokinase   \n",
              "PGI                           Glucose-6-phosphate isomerase   \n",
              "PGK                                 Phosphoglycerate kinase   \n",
              "PGM                                 Phosphoglycerate mutase   \n",
              "PYK                                         Pyruvate kinase   \n",
              "TPI                              Triose-phosphate isomerase   \n",
              "DXPS                1-deoxy-D-xylulose 5-phosphate synthase   \n",
              "DXPRIi                  1-deoxy-D-xylulose reductoisomerase   \n",
              "MEPCT     2-C-methyl-D-erythritol 4-phosphate cytidylylt...   \n",
              "CDPMEK    4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...   \n",
              "MECDPS    2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...   \n",
              "MECDPDH5  2C-methyl-D-erythritol 2,4 cyclodiphosphate de...   \n",
              "IPDPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...   \n",
              "DMPPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...   \n",
              "IPDDI                   Isopentenyl-diphosphate D-isomerase   \n",
              "DMATT                         Dimethylallyltranstransferase   \n",
              "GRTT                                Geranyltranstransferase   \n",
              "GDS                     geranylgeranyl diphosphate synthase   \n",
              "CPS                                15-cis-phytoene synthase   \n",
              "PDS                                     phytoene desaturase   \n",
              "\n",
              "                                         Reaction_subsystem  \\\n",
              "ENO          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "FBA          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "GAPD         [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "HEX1         [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "PFK          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "PGI          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "PGK          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "PGM          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "PYK          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "TPI          [Glycolysis/Gluconeogenesis, Lycopene pathway]   \n",
              "DXPS      [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "DXPRIi    [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "MEPCT     [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "CDPMEK    [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "MECDPS    [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "MECDPDH5  [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "IPDPS     [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "DMPPS     [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "IPDDI     [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "DMATT     [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "GRTT      [Cofactor and Prosthetic Group Biosynthesis, L...   \n",
              "GDS                                        Lycopene pathway   \n",
              "CPS                                        Lycopene pathway   \n",
              "PDS                                        Lycopene pathway   \n",
              "\n",
              "                                          Reaction_Formulas  \n",
              "ENO                                 2pg_c <=> h2o_c + pep_c  \n",
              "FBA                                fdp_c <=> dhap_c + g3p_c  \n",
              "GAPD        g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c  \n",
              "HEX1               atp_c + glc__D_c --> adp_c + g6p_c + h_c  \n",
              "PFK                   atp_c + f6p_c --> adp_c + fdp_c + h_c  \n",
              "PGI                                         g6p_c <=> f6p_c  \n",
              "PGK                       3pg_c + atp_c <=> 13dpg_c + adp_c  \n",
              "PGM                                         2pg_c <=> 3pg_c  \n",
              "PYK                   adp_c + h_c + pep_c --> atp_c + pyr_c  \n",
              "TPI                                        dhap_c <=> g3p_c  \n",
              "DXPS               g3p_c + h_c + pyr_c --> co2_c + dxyl5p_c  \n",
              "DXPRIi        dxyl5p_c + h_c + nadph_c --> 2me4p_c + nadp_c  \n",
              "MEPCT             2me4p_c + ctp_c + h_c --> 4c2me_c + ppi_c  \n",
              "CDPMEK          4c2me_c + atp_c --> 2p4c2me_c + adp_c + h_c  \n",
              "MECDPS                       2p4c2me_c --> 2mecdp_c + cmp_c  \n",
              "MECDPDH5  2mecdp_c + 2.0 flxr_c + h_c --> 2.0 flxso_c + ...  \n",
              "IPDPS     h2mb4p_c + h_c + nadh_c --> h2o_c + ipdp_c + n...  \n",
              "DMPPS     h2mb4p_c + h_c + nadh_c --> dmpp_c + h2o_c + n...  \n",
              "IPDDI                                     ipdp_c <=> dmpp_c  \n",
              "DMATT                    dmpp_c + ipdp_c --> grdp_c + ppi_c  \n",
              "GRTT                     grdp_c + ipdp_c --> frdp_c + ppi_c  \n",
              "GDS                      frdp_c + ipdp_c --> ggdp_c + ppi_c  \n",
              "CPS                           2 ggdp_c --> 15cp_c + 2 ppi_c  \n",
              "PDS                  15cp_c + 4 fad_c --> 4 fadh2_c + lcp_c  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-cceb8a87-7b4f-4a80-82f0-e73205535a91\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Reaction_Names</th>\n",
              "      <th>Reaction_subsystem</th>\n",
              "      <th>Reaction_Formulas</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENO</th>\n",
              "      <td>Enolase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>2pg_c &lt;=&gt; h2o_c + pep_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>FBA</th>\n",
              "      <td>Fructose-bisphosphate aldolase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>fdp_c &lt;=&gt; dhap_c + g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GAPD</th>\n",
              "      <td>Glyceraldehyde-3-phosphate dehydrogenase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>g3p_c + nad_c + pi_c &lt;=&gt; 13dpg_c + h_c + nadh_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HEX1</th>\n",
              "      <td>Hexokinase (D-glucose:ATP)</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>atp_c + glc__D_c --&gt; adp_c + g6p_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PFK</th>\n",
              "      <td>Phosphofructokinase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>atp_c + f6p_c --&gt; adp_c + fdp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGI</th>\n",
              "      <td>Glucose-6-phosphate isomerase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>g6p_c &lt;=&gt; f6p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGK</th>\n",
              "      <td>Phosphoglycerate kinase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>3pg_c + atp_c &lt;=&gt; 13dpg_c + adp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGM</th>\n",
              "      <td>Phosphoglycerate mutase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>2pg_c &lt;=&gt; 3pg_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PYK</th>\n",
              "      <td>Pyruvate kinase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>adp_c + h_c + pep_c --&gt; atp_c + pyr_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TPI</th>\n",
              "      <td>Triose-phosphate isomerase</td>\n",
              "      <td>[Glycolysis/Gluconeogenesis, Lycopene pathway]</td>\n",
              "      <td>dhap_c &lt;=&gt; g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DXPS</th>\n",
              "      <td>1-deoxy-D-xylulose 5-phosphate synthase</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>g3p_c + h_c + pyr_c --&gt; co2_c + dxyl5p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DXPRIi</th>\n",
              "      <td>1-deoxy-D-xylulose reductoisomerase</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>dxyl5p_c + h_c + nadph_c --&gt; 2me4p_c + nadp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MEPCT</th>\n",
              "      <td>2-C-methyl-D-erythritol 4-phosphate cytidylylt...</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>2me4p_c + ctp_c + h_c --&gt; 4c2me_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CDPMEK</th>\n",
              "      <td>4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>4c2me_c + atp_c --&gt; 2p4c2me_c + adp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPS</th>\n",
              "      <td>2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>2p4c2me_c --&gt; 2mecdp_c + cmp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPDH5</th>\n",
              "      <td>2C-methyl-D-erythritol 2,4 cyclodiphosphate de...</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>2mecdp_c + 2.0 flxr_c + h_c --&gt; 2.0 flxso_c + ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; h2o_c + ipdp_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMPPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; dmpp_c + h2o_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDDI</th>\n",
              "      <td>Isopentenyl-diphosphate D-isomerase</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>ipdp_c &lt;=&gt; dmpp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMATT</th>\n",
              "      <td>Dimethylallyltranstransferase</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>dmpp_c + ipdp_c --&gt; grdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GRTT</th>\n",
              "      <td>Geranyltranstransferase</td>\n",
              "      <td>[Cofactor and Prosthetic Group Biosynthesis, L...</td>\n",
              "      <td>grdp_c + ipdp_c --&gt; frdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GDS</th>\n",
              "      <td>geranylgeranyl diphosphate synthase</td>\n",
              "      <td>Lycopene pathway</td>\n",
              "      <td>frdp_c + ipdp_c --&gt; ggdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CPS</th>\n",
              "      <td>15-cis-phytoene synthase</td>\n",
              "      <td>Lycopene pathway</td>\n",
              "      <td>2 ggdp_c --&gt; 15cp_c + 2 ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PDS</th>\n",
              "      <td>phytoene desaturase</td>\n",
              "      <td>Lycopene pathway</td>\n",
              "      <td>15cp_c + 4 fad_c --&gt; 4 fadh2_c + lcp_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-cceb8a87-7b4f-4a80-82f0-e73205535a91')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-cceb8a87-7b4f-4a80-82f0-e73205535a91 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-cceb8a87-7b4f-4a80-82f0-e73205535a91');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 343
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Since BiGG model analysis is based on the Secretion and Uptake, We need to add two metabolites and three reactions (actually not reaction, more like a membrane transfer) to our model. <br>\n",
        "lycopene (cytosol) -> lycopene (periplasmic) -> lycopene (extracellular)"
      ],
      "metadata": {
        "id": "KJoUBBKpM8fO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# as usual metabolites first\n",
        "lcp_p = Metabolite('lcp_p', name= 'All-trans-lycopene', formula = 'C40H56', \n",
        "                   compartment= 'p')\n",
        "lcp_e = Metabolite('lcp_e', name= 'All-trans-lycopene', formula = 'C40H56', \n",
        "                   compartment= 'e')\n",
        "ecoli_model.add_metabolites([lcp_p,lcp_e])\n",
        "\n",
        "#check your work\n",
        "ecoli_model.metabolites.lcp_p"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "miyWP8u6OSwH",
        "outputId": "875cb88a-c045-49ea-e7ac-c446ac5023d1"
      },
      "execution_count": 344,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Metabolite lcp_p at 0x7f5a0bb9f9d0>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Metabolite identifier</strong></td><td>lcp_p</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>All-trans-lycopene</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0bb9f9d0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Formula</strong></td><td>C40H56</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Compartment</strong></td><td>p</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>In 0 reaction(s)</strong></td><td>\n",
              "                    \n",
              "                    </td>\n",
              "            </tr>\n",
              "        </table>"
            ]
          },
          "metadata": {},
          "execution_count": 344
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "ecoli_model.metabolites.lcp_e"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 142
        },
        "id": "AOGeGObsfhn7",
        "outputId": "eac9c627-4bfc-4402-fe39-97637017a166"
      },
      "execution_count": 345,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Metabolite lcp_e at 0x7f5a0bb9f490>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Metabolite identifier</strong></td><td>lcp_e</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>All-trans-lycopene</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0bb9f490</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Formula</strong></td><td>C40H56</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Compartment</strong></td><td>e</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>In 0 reaction(s)</strong></td><td>\n",
              "                    \n",
              "                    </td>\n",
              "            </tr>\n",
              "        </table>"
            ]
          },
          "metadata": {},
          "execution_count": 345
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "lcp_c = ecoli_model.metabolites.lcp_c\n",
        "\n",
        "reaction_4 = Reaction('EX_lcp_e', name = 'Lycopene exchange', lower_bound = 0, upper_bound=1000) \n",
        "reaction_4.add_metabolites({lcp_e: -1})\n",
        "# there might be intermembrane protein or other type of protein that could be effect transport\n",
        "# but I will leave GPR empty.\n",
        "\n",
        "reaction_5 = Reaction('LCPtex', name= 'Lycopene transport via diffusion (extracellular to periplasm)', lower_bound= -1000, upper_bound= 1000) # since it is transfer,same lower and upper bound.\n",
        "reaction_5.add_metabolites({lcp_e: -1, lcp_p: 1})\n",
        "\n",
        "reaction_6 = Reaction('LCPtrpp', name= 'Lycopene reversible transport via diffusion (periplasm)', lower_bound= -1000, upper_bound= 1000)\n",
        "reaction_6.add_metabolites({lcp_p: -1, lycopene: 1})\n",
        "\n",
        "ecoli_model.add_reactions([reaction_4, reaction_5, reaction_6])\n",
        "ecoli_model.reactions.get_by_id('EX_lcp_e')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "W4IC3_vGfnX7",
        "outputId": "767570ce-9c5c-40f2-8ec6-21d1a62dfbc3"
      },
      "execution_count": 346,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction EX_lcp_e at 0x7f5a0bbaa0a0>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>EX_lcp_e</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>Lycopene exchange</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0bbaa0a0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>lcp_e --></p>\n",
              "                    <p style='text-align:right'>All-trans-lycopene --></p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td></td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 346
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "ecoli_model.reactions.get_by_id('LCPtex')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "LGNgtDtRqF7y",
        "outputId": "3e3b4295-6bfc-4388-863c-b38220ae72d6"
      },
      "execution_count": 347,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction LCPtex at 0x7f5a0bbaa1c0>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>LCPtex</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>Lycopene transport via diffusion (extracellular to periplasm)</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0bbaa1c0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>lcp_e <=> lcp_p</p>\n",
              "                    <p style='text-align:right'>All-trans-lycopene <=> All-trans-lycopene</p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td></td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>-1000</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 347
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "ecoli_model.reactions.get_by_id('LCPtrpp')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "2gOitL0_qNtm",
        "outputId": "62fbbff2-887a-4e04-f775-690e67d327e0"
      },
      "execution_count": 348,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction LCPtrpp at 0x7f5a0bbaa190>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>LCPtrpp</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>Lycopene reversible transport via diffusion (periplasm)</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0bbaa190</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>lcp_p <=> lcp_c</p>\n",
              "                    <p style='text-align:right'>All-trans-lycopene <=> All-trans-lycopene</p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td></td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>-1000</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 348
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Lets check that our pathway and secretion reactions are in the system by running <i>ecoli_model.summary()"
      ],
      "metadata": {
        "id": "fxtumAL87kZK"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "ecoli_model.optimize()\n",
        "ecoli_model.summary()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 974
        },
        "id": "z34muWNvgYRg",
        "outputId": "2bdaad08-15d2-4782-ce32-24a04cbf7f8e"
      },
      "execution_count": 368,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<cobra.summary.model_summary.ModelSummary at 0x7f5a0ea26520>"
            ],
            "text/html": [
              "<h3>Objective</h3><p>1.0 BIOMASS_Ec_iJO1366_core_53p95M = 0.3389230615965925</p><h4>Uptake</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>ca2_e</td>\n",
              "      <td>EX_ca2_e</td>\n",
              "      <td>0.001764</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cl_e</td>\n",
              "      <td>EX_cl_e</td>\n",
              "      <td>0.001764</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cobalt2_e</td>\n",
              "      <td>EX_cobalt2_e</td>\n",
              "      <td>8.473E-06</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>cu2_e</td>\n",
              "      <td>EX_cu2_e</td>\n",
              "      <td>0.0002403</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>fe2_e</td>\n",
              "      <td>EX_fe2_e</td>\n",
              "      <td>0.005443</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>glc__D_e</td>\n",
              "      <td>EX_glc__D_e</td>\n",
              "      <td>13.5</td>\n",
              "      <td>6</td>\n",
              "      <td>100.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>k_e</td>\n",
              "      <td>EX_k_e</td>\n",
              "      <td>0.06616</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>mg2_e</td>\n",
              "      <td>EX_mg2_e</td>\n",
              "      <td>0.00294</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>mn2_e</td>\n",
              "      <td>EX_mn2_e</td>\n",
              "      <td>0.0002342</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>mobd_e</td>\n",
              "      <td>EX_mobd_e</td>\n",
              "      <td>4.372E-05</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>nh4_e</td>\n",
              "      <td>EX_nh4_e</td>\n",
              "      <td>3.661</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>ni2_e</td>\n",
              "      <td>EX_ni2_e</td>\n",
              "      <td>0.0001095</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>o2_e</td>\n",
              "      <td>EX_o2_e</td>\n",
              "      <td>11.46</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>pi_e</td>\n",
              "      <td>EX_pi_e</td>\n",
              "      <td>0.3269</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>so4_e</td>\n",
              "      <td>EX_so4_e</td>\n",
              "      <td>0.08548</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>zn2_e</td>\n",
              "      <td>EX_zn2_e</td>\n",
              "      <td>0.0001156</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table><h4>Secretion</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Metabolite</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Flux</th>\n",
              "      <th>C-Number</th>\n",
              "      <th>C-Flux</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>4crsol_c</td>\n",
              "      <td>DM_4crsol_c</td>\n",
              "      <td>-7.558E-05</td>\n",
              "      <td>7</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>5drib_c</td>\n",
              "      <td>DM_5drib_c</td>\n",
              "      <td>-7.626E-05</td>\n",
              "      <td>5</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>amob_c</td>\n",
              "      <td>DM_amob_c</td>\n",
              "      <td>-6.778E-07</td>\n",
              "      <td>15</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>mththf_c</td>\n",
              "      <td>DM_mththf_c</td>\n",
              "      <td>-0.0001518</td>\n",
              "      <td>5</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>co2_e</td>\n",
              "      <td>EX_co2_e</td>\n",
              "      <td>-26.42</td>\n",
              "      <td>1</td>\n",
              "      <td>99.99%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>h2o_e</td>\n",
              "      <td>EX_h2o_e</td>\n",
              "      <td>-47.57</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>h_e</td>\n",
              "      <td>EX_h_e</td>\n",
              "      <td>-3.114</td>\n",
              "      <td>0</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <td>meoh_e</td>\n",
              "      <td>EX_meoh_e</td>\n",
              "      <td>-6.778E-07</td>\n",
              "      <td>1</td>\n",
              "      <td>0.00%</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>"
            ]
          },
          "metadata": {},
          "execution_count": 368
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Oh no, even tough we added all metabolites and reactions, our lycopene is still not here. Why? Because cobra thinks that your reactions are not important enough to be included as part of the objective reactions. In order to see the lycopene flux in the summary list, we need to add our reactions to the objective reactions.\n",
        "\n",
        "What is objective reactions? In cobrapy, the objective reactions are a subset of the reactions in a metabolic model that are used to define the objective function. The objective function is a mathematical expression that represents the optimization goal of the model, such as maximizing biomass production, ATP yield, or a specific metabolic flux.\n",
        "\n",
        "by ChatGPT"
      ],
      "metadata": {
        "id": "AMoAnqF4gegS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Objective function for optimal growth\n",
        "ecoli_model.reactions.get_by_id('BIOMASS_Ec_iJO1366_core_53p95M')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 198
        },
        "id": "kgshffQHaJ1e",
        "outputId": "14ed7273-5b06-4af4-80f5-a5bf3fc2558b"
      },
      "execution_count": 350,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Reaction BIOMASS_Ec_iJO1366_core_53p95M at 0x7f5a0ce09490>"
            ],
            "text/html": [
              "\n",
              "        <table>\n",
              "            <tr>\n",
              "                <td><strong>Reaction identifier</strong></td><td>BIOMASS_Ec_iJO1366_core_53p95M</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Name</strong></td><td>E. coli biomass objective function (iJO1366) - core - with 53.95 GAM estimate</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Memory address</strong></td>\n",
              "                <td>0x7f5a0ce09490</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Stoichiometry</strong></td>\n",
              "                <td>\n",
              "                    <p style='text-align:right'>0.000223 10fthf_c + 2.6e-05 2fe2s_c + 0.000223 2ohph_c + 0.00026 4fe4s_c + 0.513689 ala__L_c + 0.000223 amet_c + 0.295792 arg__L_c + 0.241055 asn__L_c + 0.241055 asp__L_c + 54.124831 atp_c +...</p>\n",
              "                    <p style='text-align:right'>0.000223 10-Formyltetrahydrofolate + 2.6e-05 [2Fe-2S] iron-sulfur cluster + 0.000223 2-Octaprenyl-6-hydroxyphenol + 0.00026 [4Fe-4S] iron-sulfur cluster + 0.513689 L-Alanine + 0.000223 S-Adenosyl-...</p>\n",
              "                </td>\n",
              "            </tr><tr>\n",
              "                <td><strong>GPR</strong></td><td></td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
              "            </tr><tr>\n",
              "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
              "            </tr>\n",
              "        </table>\n",
              "        "
            ]
          },
          "metadata": {},
          "execution_count": 350
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        " if you want to optimize the production of lycopene, you can modify the biomass objective function as follows"
      ],
      "metadata": {
        "id": "5L-zXg2jb1sN"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# add lycopene as a product with a stoichiometric coefficient of -1.0\n",
        "ecoli_model.reactions.get_by_id('BIOMASS_Ec_iJO1366_core_53p95M').add_metabolites({ecoli_model.metabolites.lcp_e: -1})\n",
        "\n",
        "# Set the objective coefficient of the lyctopene production reaction to 1.0\n",
        "ecoli_model.reactions.get_by_id('BIOMASS_Ec_iJO1366_core_53p95M').objective_coefficient = 1.0"
      ],
      "metadata": {
        "id": "0bOaGz3DcWAH"
      },
      "execution_count": 354,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "This modification will set the acetate production as the objective of the optimization problem, while still ensuring that the biomass precursors are synthesized in sufficient quantities for growth. You can then use FBA to find the optimal flux distribution for all reactions in the metabolic network under the constraint of maximum lycopene production."
      ],
      "metadata": {
        "id": "fytoKi5icWtD"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from cobra.util.solver import linear_reaction_coefficients\n",
        "\n",
        "# Set the maximum lycopene production as the constraint\n",
        "ecoli_model.reactions.EX_lcp_e.upper_bound = 10.0  # set an upper bound of 10 mmol/gDW/hr for lycopene exchange\n",
        "\n",
        "linear_reaction_coefficients(ecoli_model)\n",
        "ecoli_model.optimize()\n",
        "ecoli_model.metabolites.lcp_e.summary()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "KC6GaFfPUcgH",
        "outputId": "926fc376-faf2-45d4-bdd8-688a723656ec"
      },
      "execution_count": 356,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<cobra.summary.metabolite_summary.MetaboliteSummary at 0x7f5a0bb9fe80>"
            ],
            "text/html": [
              "<h3>lcp_e</h3><p>C40H56</p><h4>Producing Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Percent</th>\n",
              "      <th>Flux</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Definition</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>100.00%</td>\n",
              "      <td>0.7507</td>\n",
              "      <td>LCPtex</td>\n",
              "      <td>lcp_e &lt;=&gt; lcp_p</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table><h4>Consuming Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Percent</th>\n",
              "      <th>Flux</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Definition</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>100.00%</td>\n",
              "      <td>-0.7507</td>\n",
              "      <td>BIOMASS_Ec_iJO1366_core_53p95M</td>\n",
              "      <td>0.000223 10fthf_c + 2.6e-05 2fe2s_c + 0.000223 2ohph_c + 0.00026 4fe4s_c + 0.513689 ala__L_c + 0.000223 amet_c + 0.295792 arg__L_c + 0.241055 asn__L_c + 0.241055 asp__L_c + 54.124831 atp_c + 0.000122 bmocogdp_c + 2e-06 btn_c + 0.005205 ca2_c + 0.005205 cl_c + 0.000576 coa_c + 2.5e-05 cobalt2_c + 0.133508 ctp_c + 0.000709 cu2_c + 0.09158 cys__L_c + 0.026166 datp_c + 0.027017 dctp_c + 0.027017 dgtp_c + 0.026166 dttp_c + 0.000223 fad_c + 0.006715 fe2_c + 0.007808 fe3_c + 0.26316 gln__L_c + 0.26316 glu__L_c + 0.612638 gly_c + 0.215096 gtp_c + 48.601527 h2o_c + 0.094738 his__L_c + 0.290529 ile__L_c + 0.195193 k_c + 0.019456 kdo2lipid4_e + 3 lcp_e + 0.450531 leu__L_c + 0.343161 lys__L_c + 0.153686 met__L_c + 0.008675 mg2_c + 0.000223 mlthf_c + 0.000691 mn2_c + 7e-06 mobd_c + 0.013894 murein5px4p_p + 0.001831 nad_c + 0.000447 nadp_c + 0.013013 nh4_c + 0.000323 ni2_c + 0.017868 pe160_c + 0.045946 pe160_p + 0.054154 pe161_c + 0.02106 pe161_p + 0.185265 phe__L_c + 0.000223 pheme_c + 0.221055 pro__L_c + 0.000223 pydx5p_c + 0.000223 ribflv_c + 0.215792 ser__L_c + 0.000223 sheme_c + 0.004338 so4_c + 0.000223 thf_c + 0.000223 thmpp_c + 0.253687 thr__L_c + 0.056843 trp__L_c + 0.137896 tyr__L_c + 5.5e-05 udcpdp_c + 0.144104 utp_c + 0.423162 val__L_c + 0.000341 zn2_c --&gt; 53.95 adp_c + 53.95 h_c + 53.945662 pi_c + 0.773903 ppi_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>"
            ]
          },
          "metadata": {},
          "execution_count": 356
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "This code will optimize the flux distribution in the network to maximize the production of lycopene, subject to the constraint that the maximum exchange flux for lycopene is 10 mmol/gDW/hr. The output of the optimization will be the optimal flux through the lycopene exchange reaction, which represents the maximum lycopene production rate that the network can support under the given constraint.\n",
        "\n",
        "This objective supports both max growth and production of lycopene."
      ],
      "metadata": {
        "id": "PADIDEoMem2Z"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let see how it will change by changing the flux of the glucose input"
      ],
      "metadata": {
        "id": "_HzuiGeSna4B"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "ecoli_model.reactions.EX_glc__D_e.lower_bound = -20\n",
        "\n",
        "ecoli_model.optimize()\n",
        "ecoli_model.metabolites.lcp_e.summary()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "id": "zoDlSyGYk4zK",
        "outputId": "ec6de333-7f92-48f3-8919-adb81630fa55"
      },
      "execution_count": 358,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<cobra.summary.metabolite_summary.MetaboliteSummary at 0x7f5a0d1d3f10>"
            ],
            "text/html": [
              "<h3>lcp_e</h3><p>C40H56</p><h4>Producing Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Percent</th>\n",
              "      <th>Flux</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Definition</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>100.00%</td>\n",
              "      <td>1.511</td>\n",
              "      <td>LCPtex</td>\n",
              "      <td>lcp_e &lt;=&gt; lcp_p</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table><h4>Consuming Reactions</h4><table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th>Percent</th>\n",
              "      <th>Flux</th>\n",
              "      <th>Reaction</th>\n",
              "      <th>Definition</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <td>100.00%</td>\n",
              "      <td>-1.511</td>\n",
              "      <td>BIOMASS_Ec_iJO1366_core_53p95M</td>\n",
              "      <td>0.000223 10fthf_c + 2.6e-05 2fe2s_c + 0.000223 2ohph_c + 0.00026 4fe4s_c + 0.513689 ala__L_c + 0.000223 amet_c + 0.295792 arg__L_c + 0.241055 asn__L_c + 0.241055 asp__L_c + 54.124831 atp_c + 0.000122 bmocogdp_c + 2e-06 btn_c + 0.005205 ca2_c + 0.005205 cl_c + 0.000576 coa_c + 2.5e-05 cobalt2_c + 0.133508 ctp_c + 0.000709 cu2_c + 0.09158 cys__L_c + 0.026166 datp_c + 0.027017 dctp_c + 0.027017 dgtp_c + 0.026166 dttp_c + 0.000223 fad_c + 0.006715 fe2_c + 0.007808 fe3_c + 0.26316 gln__L_c + 0.26316 glu__L_c + 0.612638 gly_c + 0.215096 gtp_c + 48.601527 h2o_c + 0.094738 his__L_c + 0.290529 ile__L_c + 0.195193 k_c + 0.019456 kdo2lipid4_e + 3 lcp_e + 0.450531 leu__L_c + 0.343161 lys__L_c + 0.153686 met__L_c + 0.008675 mg2_c + 0.000223 mlthf_c + 0.000691 mn2_c + 7e-06 mobd_c + 0.013894 murein5px4p_p + 0.001831 nad_c + 0.000447 nadp_c + 0.013013 nh4_c + 0.000323 ni2_c + 0.017868 pe160_c + 0.045946 pe160_p + 0.054154 pe161_c + 0.02106 pe161_p + 0.185265 phe__L_c + 0.000223 pheme_c + 0.221055 pro__L_c + 0.000223 pydx5p_c + 0.000223 ribflv_c + 0.215792 ser__L_c + 0.000223 sheme_c + 0.004338 so4_c + 0.000223 thf_c + 0.000223 thmpp_c + 0.253687 thr__L_c + 0.056843 trp__L_c + 0.137896 tyr__L_c + 5.5e-05 udcpdp_c + 0.144104 utp_c + 0.423162 val__L_c + 0.000341 zn2_c --&gt; 53.95 adp_c + 53.95 h_c + 53.945662 pi_c + 0.773903 ppi_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>"
            ]
          },
          "metadata": {},
          "execution_count": 358
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "As we can see that higher glucose flux returns higher flux of lycopene.\n",
        "\n",
        "From the results so far, it appears that our work is yielding positive outcomes.\n",
        "\n",
        "but question is 'is it really working?'"
      ],
      "metadata": {
        "id": "NjaHWP2BoeSG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "ecoli_model.reactions.EX_glc__D_e.lower_bound = -13.5\n",
        "\n",
        "solution = ecoli_model.optimize()\n",
        "reactions_id = ['HEX1','PGI','PFK','FBA','TPI','GAPD','PGK','PGM','ENO','PYK','DXPS','DXPRIi','MEPCT','CDPMEK','MECDPS','MECDPDH5','IPDPS','DMPPS','IPDDI','DMATT','GRTT','GDS','CPS','PDS']\n",
        "\n",
        "reactions_lycophene_pathway = []\n",
        "for id in reactions_id:\n",
        "  reactions_lycophene_pathway.append(ecoli_model.reactions.get_by_id(id))\n",
        "\n",
        "reactionNames = [r.name for r in reactions_lycophene_pathway]\n",
        "reactionFormulas = [r.build_reaction_string() for r in reactions_lycophene_pathway]\n",
        "reactionFlus = [solution.fluxes[r.id]for r in reactions_lycophene_pathway]\n",
        "# take out the useless carbon atom.\n",
        "out = ['atp_c','adp_c','h_c','pi_c','nad_c','nadh_c','h2o_c','nadph_c','nadp_c','ctp_c','ppi_c','cmp_c','flxr_c','flxso_c','fad_c','fadh2_c']\n",
        "carbon_count = [[mtb.elements.get('C', 0) for mtb in r.metabolites if mtb.id not in out] for r in reactions_lycophene_pathway]\n",
        "moles_of_product = [10,10,10,20,20,20,20,20,20,20,10,10,10,10,10,10,10,10,10,[2,6],[2,4],[2,2],[1,2],[1,2]]\n",
        "carbon_numbers = [6,6,6,3,3,3,3,3,3,3,5,5,5,5,5,5,5,5,5,[10,5],[15,5],[20,5],[40,5],[40,5]]\n",
        "\n",
        "tableData = {'Reaction_Names': reactionNames,'Carbon number': carbon_count, 'Moles': moles_of_product, 'Reaction_Flux' : reactionFlus, 'Reaction_Formulas': reactionFormulas}\n",
        "tableIndex = [r.id for r in reactions_lycophene_pathway]\n",
        "T = pd.DataFrame(tableData, index=tableIndex)\n",
        "T"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 802
        },
        "id": "gHJSy4ZpqYnO",
        "outputId": "efdb6332-3da7-417c-f649-5e6ab3c84c8f"
      },
      "execution_count": 365,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                                             Reaction_Names Carbon number  \\\n",
              "HEX1                             Hexokinase (D-glucose:ATP)        [6, 6]   \n",
              "PGI                           Glucose-6-phosphate isomerase        [6, 6]   \n",
              "PFK                                     Phosphofructokinase        [6, 6]   \n",
              "FBA                          Fructose-bisphosphate aldolase     [6, 3, 3]   \n",
              "TPI                              Triose-phosphate isomerase        [3, 3]   \n",
              "GAPD               Glyceraldehyde-3-phosphate dehydrogenase        [3, 3]   \n",
              "PGK                                 Phosphoglycerate kinase        [3, 3]   \n",
              "PGM                                 Phosphoglycerate mutase        [3, 3]   \n",
              "ENO                                                 Enolase        [3, 3]   \n",
              "PYK                                         Pyruvate kinase        [3, 3]   \n",
              "DXPS                1-deoxy-D-xylulose 5-phosphate synthase  [3, 3, 1, 5]   \n",
              "DXPRIi                  1-deoxy-D-xylulose reductoisomerase        [5, 5]   \n",
              "MEPCT     2-C-methyl-D-erythritol 4-phosphate cytidylylt...       [5, 14]   \n",
              "CDPMEK    4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...      [14, 14]   \n",
              "MECDPS    2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...       [14, 5]   \n",
              "MECDPDH5  2C-methyl-D-erythritol 2,4 cyclodiphosphate de...        [5, 5]   \n",
              "IPDPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...        [5, 5]   \n",
              "DMPPS     1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...        [5, 5]   \n",
              "IPDDI                   Isopentenyl-diphosphate D-isomerase        [5, 5]   \n",
              "DMATT                         Dimethylallyltranstransferase    [5, 5, 10]   \n",
              "GRTT                                Geranyltranstransferase   [10, 5, 15]   \n",
              "GDS                     geranylgeranyl diphosphate synthase   [15, 5, 20]   \n",
              "CPS                                15-cis-phytoene synthase      [20, 40]   \n",
              "PDS                                     phytoene desaturase      [40, 40]   \n",
              "\n",
              "           Moles  Reaction_Flux  \\\n",
              "HEX1          10       0.000000   \n",
              "PGI           10       7.699932   \n",
              "PFK           10       9.518182   \n",
              "FBA           20       9.518182   \n",
              "TPI           20      11.223715   \n",
              "GAPD          20      16.116699   \n",
              "PGK           20     -16.116699   \n",
              "PGM           20     -15.534370   \n",
              "ENO           20      15.534370   \n",
              "PYK           20       0.000000   \n",
              "DXPS          10       8.135114   \n",
              "DXPRIi        10       8.134963   \n",
              "MEPCT         10       8.134963   \n",
              "CDPMEK        10       8.134963   \n",
              "MECDPS        10       8.134963   \n",
              "MECDPDH5      10       8.134963   \n",
              "IPDPS         10       6.101331   \n",
              "DMPPS         10       2.033633   \n",
              "IPDDI         10       0.000000   \n",
              "DMATT     [2, 6]       2.033633   \n",
              "GRTT      [2, 4]       2.033633   \n",
              "GDS       [2, 2]       2.033538   \n",
              "CPS       [1, 2]       1.016769   \n",
              "PDS       [1, 2]       1.016769   \n",
              "\n",
              "                                          Reaction_Formulas  \n",
              "HEX1               atp_c + glc__D_c --> adp_c + g6p_c + h_c  \n",
              "PGI                                         g6p_c <=> f6p_c  \n",
              "PFK                   atp_c + f6p_c --> adp_c + fdp_c + h_c  \n",
              "FBA                                fdp_c <=> dhap_c + g3p_c  \n",
              "TPI                                        dhap_c <=> g3p_c  \n",
              "GAPD        g3p_c + nad_c + pi_c <=> 13dpg_c + h_c + nadh_c  \n",
              "PGK                       3pg_c + atp_c <=> 13dpg_c + adp_c  \n",
              "PGM                                         2pg_c <=> 3pg_c  \n",
              "ENO                                 2pg_c <=> h2o_c + pep_c  \n",
              "PYK                   adp_c + h_c + pep_c --> atp_c + pyr_c  \n",
              "DXPS               g3p_c + h_c + pyr_c --> co2_c + dxyl5p_c  \n",
              "DXPRIi        dxyl5p_c + h_c + nadph_c --> 2me4p_c + nadp_c  \n",
              "MEPCT             2me4p_c + ctp_c + h_c --> 4c2me_c + ppi_c  \n",
              "CDPMEK          4c2me_c + atp_c --> 2p4c2me_c + adp_c + h_c  \n",
              "MECDPS                       2p4c2me_c --> 2mecdp_c + cmp_c  \n",
              "MECDPDH5  2mecdp_c + 2.0 flxr_c + h_c --> 2.0 flxso_c + ...  \n",
              "IPDPS     h2mb4p_c + h_c + nadh_c --> h2o_c + ipdp_c + n...  \n",
              "DMPPS     h2mb4p_c + h_c + nadh_c --> dmpp_c + h2o_c + n...  \n",
              "IPDDI                                     ipdp_c <=> dmpp_c  \n",
              "DMATT                    dmpp_c + ipdp_c --> grdp_c + ppi_c  \n",
              "GRTT                     grdp_c + ipdp_c --> frdp_c + ppi_c  \n",
              "GDS                      frdp_c + ipdp_c --> ggdp_c + ppi_c  \n",
              "CPS                           2 ggdp_c --> 15cp_c + 2 ppi_c  \n",
              "PDS                  15cp_c + 4 fad_c --> 4 fadh2_c + lcp_c  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-805532f9-9480-4af2-a0a3-d6832214a515\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Reaction_Names</th>\n",
              "      <th>Carbon number</th>\n",
              "      <th>Moles</th>\n",
              "      <th>Reaction_Flux</th>\n",
              "      <th>Reaction_Formulas</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>HEX1</th>\n",
              "      <td>Hexokinase (D-glucose:ATP)</td>\n",
              "      <td>[6, 6]</td>\n",
              "      <td>10</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>atp_c + glc__D_c --&gt; adp_c + g6p_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGI</th>\n",
              "      <td>Glucose-6-phosphate isomerase</td>\n",
              "      <td>[6, 6]</td>\n",
              "      <td>10</td>\n",
              "      <td>7.699932</td>\n",
              "      <td>g6p_c &lt;=&gt; f6p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PFK</th>\n",
              "      <td>Phosphofructokinase</td>\n",
              "      <td>[6, 6]</td>\n",
              "      <td>10</td>\n",
              "      <td>9.518182</td>\n",
              "      <td>atp_c + f6p_c --&gt; adp_c + fdp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>FBA</th>\n",
              "      <td>Fructose-bisphosphate aldolase</td>\n",
              "      <td>[6, 3, 3]</td>\n",
              "      <td>20</td>\n",
              "      <td>9.518182</td>\n",
              "      <td>fdp_c &lt;=&gt; dhap_c + g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>TPI</th>\n",
              "      <td>Triose-phosphate isomerase</td>\n",
              "      <td>[3, 3]</td>\n",
              "      <td>20</td>\n",
              "      <td>11.223715</td>\n",
              "      <td>dhap_c &lt;=&gt; g3p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GAPD</th>\n",
              "      <td>Glyceraldehyde-3-phosphate dehydrogenase</td>\n",
              "      <td>[3, 3]</td>\n",
              "      <td>20</td>\n",
              "      <td>16.116699</td>\n",
              "      <td>g3p_c + nad_c + pi_c &lt;=&gt; 13dpg_c + h_c + nadh_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGK</th>\n",
              "      <td>Phosphoglycerate kinase</td>\n",
              "      <td>[3, 3]</td>\n",
              "      <td>20</td>\n",
              "      <td>-16.116699</td>\n",
              "      <td>3pg_c + atp_c &lt;=&gt; 13dpg_c + adp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PGM</th>\n",
              "      <td>Phosphoglycerate mutase</td>\n",
              "      <td>[3, 3]</td>\n",
              "      <td>20</td>\n",
              "      <td>-15.534370</td>\n",
              "      <td>2pg_c &lt;=&gt; 3pg_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENO</th>\n",
              "      <td>Enolase</td>\n",
              "      <td>[3, 3]</td>\n",
              "      <td>20</td>\n",
              "      <td>15.534370</td>\n",
              "      <td>2pg_c &lt;=&gt; h2o_c + pep_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PYK</th>\n",
              "      <td>Pyruvate kinase</td>\n",
              "      <td>[3, 3]</td>\n",
              "      <td>20</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>adp_c + h_c + pep_c --&gt; atp_c + pyr_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DXPS</th>\n",
              "      <td>1-deoxy-D-xylulose 5-phosphate synthase</td>\n",
              "      <td>[3, 3, 1, 5]</td>\n",
              "      <td>10</td>\n",
              "      <td>8.135114</td>\n",
              "      <td>g3p_c + h_c + pyr_c --&gt; co2_c + dxyl5p_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DXPRIi</th>\n",
              "      <td>1-deoxy-D-xylulose reductoisomerase</td>\n",
              "      <td>[5, 5]</td>\n",
              "      <td>10</td>\n",
              "      <td>8.134963</td>\n",
              "      <td>dxyl5p_c + h_c + nadph_c --&gt; 2me4p_c + nadp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MEPCT</th>\n",
              "      <td>2-C-methyl-D-erythritol 4-phosphate cytidylylt...</td>\n",
              "      <td>[5, 14]</td>\n",
              "      <td>10</td>\n",
              "      <td>8.134963</td>\n",
              "      <td>2me4p_c + ctp_c + h_c --&gt; 4c2me_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CDPMEK</th>\n",
              "      <td>4-(cytidine 5'-diphospho)-2-C-methyl-D-erythri...</td>\n",
              "      <td>[14, 14]</td>\n",
              "      <td>10</td>\n",
              "      <td>8.134963</td>\n",
              "      <td>4c2me_c + atp_c --&gt; 2p4c2me_c + adp_c + h_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPS</th>\n",
              "      <td>2-C-methyl-D-erythritol 2,4-cyclodiphosphate s...</td>\n",
              "      <td>[14, 5]</td>\n",
              "      <td>10</td>\n",
              "      <td>8.134963</td>\n",
              "      <td>2p4c2me_c --&gt; 2mecdp_c + cmp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>MECDPDH5</th>\n",
              "      <td>2C-methyl-D-erythritol 2,4 cyclodiphosphate de...</td>\n",
              "      <td>[5, 5]</td>\n",
              "      <td>10</td>\n",
              "      <td>8.134963</td>\n",
              "      <td>2mecdp_c + 2.0 flxr_c + h_c --&gt; 2.0 flxso_c + ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>[5, 5]</td>\n",
              "      <td>10</td>\n",
              "      <td>6.101331</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; h2o_c + ipdp_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMPPS</th>\n",
              "      <td>1-hydroxy-2-methyl-2-(E)-butenyl 4-diphosphate...</td>\n",
              "      <td>[5, 5]</td>\n",
              "      <td>10</td>\n",
              "      <td>2.033633</td>\n",
              "      <td>h2mb4p_c + h_c + nadh_c --&gt; dmpp_c + h2o_c + n...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>IPDDI</th>\n",
              "      <td>Isopentenyl-diphosphate D-isomerase</td>\n",
              "      <td>[5, 5]</td>\n",
              "      <td>10</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>ipdp_c &lt;=&gt; dmpp_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>DMATT</th>\n",
              "      <td>Dimethylallyltranstransferase</td>\n",
              "      <td>[5, 5, 10]</td>\n",
              "      <td>[2, 6]</td>\n",
              "      <td>2.033633</td>\n",
              "      <td>dmpp_c + ipdp_c --&gt; grdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GRTT</th>\n",
              "      <td>Geranyltranstransferase</td>\n",
              "      <td>[10, 5, 15]</td>\n",
              "      <td>[2, 4]</td>\n",
              "      <td>2.033633</td>\n",
              "      <td>grdp_c + ipdp_c --&gt; frdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>GDS</th>\n",
              "      <td>geranylgeranyl diphosphate synthase</td>\n",
              "      <td>[15, 5, 20]</td>\n",
              "      <td>[2, 2]</td>\n",
              "      <td>2.033538</td>\n",
              "      <td>frdp_c + ipdp_c --&gt; ggdp_c + ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>CPS</th>\n",
              "      <td>15-cis-phytoene synthase</td>\n",
              "      <td>[20, 40]</td>\n",
              "      <td>[1, 2]</td>\n",
              "      <td>1.016769</td>\n",
              "      <td>2 ggdp_c --&gt; 15cp_c + 2 ppi_c</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>PDS</th>\n",
              "      <td>phytoene desaturase</td>\n",
              "      <td>[40, 40]</td>\n",
              "      <td>[1, 2]</td>\n",
              "      <td>1.016769</td>\n",
              "      <td>15cp_c + 4 fad_c --&gt; 4 fadh2_c + lcp_c</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-805532f9-9480-4af2-a0a3-d6832214a515')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-805532f9-9480-4af2-a0a3-d6832214a515 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-805532f9-9480-4af2-a0a3-d6832214a515');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 365
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Briefly, it is okey to me since there could be another use of 2 moles of glucose."
      ],
      "metadata": {
        "id": "sJZOeQ7-JPpP"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "<b>Work Cited "
      ],
      "metadata": {
        "id": "buWCPukupiE1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Biocyc: https://biocyc.org/META/NEW-IMAGE?type=REACTION&object=FARNESYLTRANSTRANSFERASE-RXN<br>\n",
        "Biocyc: https://biocyc.org/META/NEW-IMAGE?type=REACTION&object=RXN-13323<br>\n",
        "Biocyc: https://biocyc.org/META/NEW-IMAGE?type=REACTION&object=RXN-12414<br>\n",
        "Cobra Toolbox Tutorials: https://opencobra.github.io/cobratoolbox/stable/tutorials/index.html<br>\n",
        "Cobrapy Tutorial: https://github.com/migp11/cobrapy-tutorial<br>\n",
        "Cobrapy Documentation - Simulating: https://cobrapy.readthedocs.io/en/latest/simulating.html<br>\n",
        "ChatGPT (2023). \"Conversation with ChatGPT on partially objective function and flux description.\""
      ],
      "metadata": {
        "id": "_8uCNYglqGm4"
      }
    }
  ]
}